GOVERNANCE.md

```diff
@@ -22,30 +18,164 @@ New maintainers must be nominated by an existing maintainer and must be elected
+please use whichever sections make the most sense for your proposal.
+
+```md
 ---
-# Updating Governance
-All substantive changes in Governance require a supermajority agreement by all maintainers.
\ No newline at end of file
+title: "The Name of My Proposal"
+authors: [ "@margocrawf", "@enj" ]
+status: "draft"
+sponsor: [ "@cfryanr" ]
+---
+
+# <Proposal Title>
+
+## Problem Statement
+This is a short summary of the problem that exists, why it needs to be
+solved: what specific needs are being met. Compelling problem statements
+include concrete examples and use cases (even if only by reference).
+How exactly the proposal would meet those needs should be located in the
+"Proposal" section, not this one. The goal of this section is to help
+readers quickly empathize with the target users' current experience to
+motivate the proposed change.
+
+### How Pinniped Works Today (as of version vX.X.X)
+How Pinniped works today in the context of the problem statement.
+This will typically detail how Pinniped falls short of supporting
+the desired use case(s).
+
+## Terminology / Concepts
+Define any terms or concepts that are used throughout this proposal.
+
+## Proposal
+The primary content of the proposal. Subsections will explain how the
+problem(s) will be addressed.
+
+### Goals and Non-goals
+A short list of what the goals of this proposal are and are not.
+
+### Specification / How it Solves the Use Cases
+Detailed explanation of the proposal's design. This will typically
+also detail how the specification supports the desired use cases.
+
+### Other Approaches Considered
+Mention of other reasonable ways that the problem(s)
+could be addressed with rationale for why they were less
+desirable than the proposed approach.
+
+## Open Questions
+A list of questions that need to be answered.
+
+## Answered Questions
+A list of questions that have been answered.
+```
+
+## Proposal States
+| Status | Definition |
+| --- | --- |
+| `draft` | The proposal is actively being written by the proposer. Not yet ready for review. |
+| `in-review` | The proposal is being reviewed by the community and the project maintainers. |
+| `accepted` | The proposal has been accepted by the project maintainers. |
+| `rejected` | The proposal has been rejected by the project maintainers. |
+
+## Lifecycle of a Proposal
+1. Author adds a proposal by creating a PR in draft mode. (Authors can save their work until ready.)
+2. When the author elaborates the proposal sufficiently to withstand critique they:
+   1. change the status to `in-review` and
+   2. mark the PR as "Ready for Review"
+3. The community critiques the proposal by adding PR reviews in order to mature/converge on the proposal.
+4. When the maintainers reach consensus or supermajority to accept a proposal, they:
+   1. change the status to `accepted`,
+   2. adjust the proposal number in the subdirectory's name if needed,
```

Comment:
It didn't feel like a proposal from the community necessarily needs to be tied to a particular GitHub issue. There could be no issue, or more likely there could be several related issues. The proposal PR itself is an issue so it can hold the conversation for historical purposes.

Maybe there's a better way to number the proposals, although I would imagine that there would be so few accepted proposals that it would not be arduous to manually number them. As a community user looking at the list of proposal subdirectories on main, I think the only thing I would care about is seeing them listed in roughly chronological order, so anything that achieves that outcome should work.

Reply:
To be clear, what I am saying is that I want the proposal to be anchored via a GitHub issue instead of a PR.  i.e. the proposal process starts with making an issue, that represents the proposal itself (and the number assigned to the issue becomes the immutable proposal number).  I expect multiple PRs to be made to the proposal over time, and for all of those PRs to refer to the proposal issue.  I also expect multiple PRs to implement a proposal, and for those to refer to the proposal issue as well.

This is one of my favorite aspects of Kube's KEP process, and I would like it replicated here because I find it to be both valuable and lightweight.

See http://kep.k8s.io/2784 as an example.

Reply:
I agree with @enj, I think using a GitHub issue to anchor the proposal process makes a lot of sense.  The lifespan of a proposal is long, longer than a single PR & likely extending over many. As pointed out above, there can be PRs to update the proposal itself in addition to PRs for the implementation.  

Having the ability to submit PRs against the proposal document seems desirable as it allows folks to collaborate more freely by proposing changes directly. This would also allow somebody else in the community to take the lead on a proposal if the initial author were to become unavailable to contribute. 

Having the ability to track implementation PRs via the issue also seems desirable. I imagine proposals potentially being implemented piecemeal over longer periods of time. I also imagine encountering PRs submitted independently that have some impact on an existing feature proposal.

Having a place where community members can easily follow the evolution of the feature proposal, design, and eventual implementation seems like a good thing.

Reply:
I don't think I understand the suggestion, @enj. I'm not very familiar with the Kubernetes process and I can't find a document which describes it very well. The main README of https://github.com/kubernetes/enhancements talks about creating issues but doesn't describe their lifecycle. They have a whole separate repo just for those issues, which feels a little heavyweight for us. As an outsider trying to guess their process, I would guess that the process is that the tracking issue is left open forever until a bot closes it as "rotten", because it seems like that's how most issues in that repo end up, e.g. https://github.com/kubernetes/enhancements/issues/2835.

Could you be more specific in what you would like to propose for our proposal process?

The process that I outlined is one in which an accepted proposal is a PR that gets merged and therefore the proposal becomes a markdown document on the main branch of our source code repo. So a future change to that proposal is simply any new PR that makes changes that markdown file on the main branch. Like any PR, that update to the proposal doc would need to get the approval of the maintainers. And like any file, git provides the complete history of the proposal doc. If a proposal needs to evolve during implementation, then a new PR could be opened which only makes changes to the proposal doc, while implementation is paused. Each implementation PR could itself also update the proposal document to update the proposal's iterative implementation status, since the code and the proposal docs are in the same repo.

Interestingly, KEP-0000 says that one of the reasons that they invented KEPs was to avoid having these umbrella issues in the first place.

> - reserve GitHub issues for tracking work in flight, instead of creating "umbrella" issues

From: https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md#summary

But then they seem to have changed their minds (without updating KEP-0000!) because the README now directly contradicts KEP-0000:

> These issues are umbrellas for new enhancements to be added to Kubernetes. 

From: https://github.com/kubernetes/enhancements#enhancement-tracking-and-backlog

Reply:
In addition to specifying the content and lifecycle of the umbrella/tracking issues in this process document, we should also discuss where we would like to put those in Tracker, since that decides what labels would get automatically applied to the GitHub issues. (This document doesn't need to talk about Tracker though, I think.)